---
title: "Distribution of GANTC sites"
output: html_document
date: "2024-02-01"
---

# Note that in order to run this script, you will require both full genome and gene sequence information for given entry from NCBI
# Libraries
```{r}
library(seqinr)
library(ggplot2)
library(dplyr)
library(stringr)

```

# Calculate Genome Statistics in relation to GANTC sites
```{r}
calculateGenomeStatistics <- function(fullGenome, geneInfo, class, order, species, strain, chromosomes) {
  genomeSeq <- paste(unlist(fullGenome), collapse = "")
  geneSeq <- paste(unlist(geneInfo), collapse = "")
  
  genomeLength <- nchar(genomeSeq)
  orfLength <- nchar(geneSeq)
  intergenicLength <- genomeLength - orfLength
  percentORF <- round(orfLength / genomeLength * 100, 3)
  
  gantcCount <- stringr::str_count(genomeSeq, "ga[atcg]tc") * 2
  orfCount <- stringr::str_count(geneSeq, "ga[atcg]tc") * 2
  intergenicCount <- gantcCount - orfCount
  
  nucleotideCounts <- table(strsplit(genomeSeq, "")[[1]])
  nucleotideFrequencies <- nucleotideCounts / genomeLength
  gcContent <- sum(nucleotideCounts[c("g", "c")]) / genomeLength
  atContent <- sum(nucleotideCounts[c("a", "t")]) / genomeLength
  
  expectedGANTC <- prod(nucleotideFrequencies) * genomeLength
  expectedGANTCGC <- gcContent * expectedGANTC
  expectedGANTCAT <- atContent * expectedGANTC
  
  countGANTCGC <- stringr::str_count(genomeSeq, "ga[gc]tc") * 2
  countGANTCAT <- stringr::str_count(genomeSeq, "ga[at]tc") * 2
  
  expectedGANTCORF <- orfLength / genomeLength * expectedGANTC
  expectedGANTCIntergenic <- intergenicLength / genomeLength * expectedGANTC
  
  observedExpectedRatio <- gantcCount / expectedGANTC
  
  df <- data.frame(
    Class = class, Order = order, Species = species, Strain = strain, 
    GenomeLength = genomeLength, Chromosomes = chromosomes, GCContent = gcContent,
    ExpectedGANTC = expectedGANTC, CountedGANTC = gantcCount,
    ExpectedGANTCAT = expectedGANTCAT, ExpectedGANTCGC = expectedGANTCGC,
    CountedGANTCAT = countGANTCAT, CountedGANTCGC = countGANTCGC,
    IntergenicLength = intergenicLength, ORFLength = orfLength,
    CountedGANTCIntergenic = intergenicCount, CountedGANTCORF = orfCount,
    ExpectedGANTCIntergenic = expectedGANTCIntergenic, ExpectedGANTCORF = expectedGANTCORF
  )
  
  return(df)
}

```


# Example
```{r}

# Full genome fasta was reference genome downloaded from NCBI 
# Gene genome .txt was gene info file downloaded from NCBI 
calculateGenomeStatistics(fullGenome = read.fasta("../../RData/Supplementary_Data/Caulobacter_crescentus_NA1000.fasta"), 
                               geneInfo = read.fasta("../../RData/Supplementary_Data/Caulobacter_crescentus_NA1000.txt"), 
                               class = "Alphaproteobacteria", 
                               order = "Caulobacterales", 
                               species = "C. Crescentus", 
                               strain = "NA_1000", 
                               chromosomes = 1)
```


```{r}
Alphaproteobacteria <-read_csv("../../RData/Supplementary_Data/Alphaproteobacteria.csv")

Alphaproteobacteria %>%
  ggplot(aes(x = Order, y = CountedGANTC / ExpectedGANTC)) +
  geom_boxplot() +
  geom_hline(yintercept = 1, linetype = "solid", color = "black") +
  theme_classic() +
  geom_point(data = Alphaproteobacteria %>% filter(Strain == "NA_1000"), 
             aes(x = Order, y = CountedGANTCAT / ExpectedGANTCAT, color = "C.crescentus NA1000"), size = 3) +
  geom_point(data = Alphaproteobacteria %>% filter(Strain == "2308"), 
             aes(x = Order, y = CountedGANTC / ExpectedGANTC,  color= paste("Brucella 2308","Chromosome: ", Chromosomes)), size = 3) +
  ylab("Counted / Expected # of GANTC site") + 
  theme(legend.position = "none") + 
  ggtitle("Genome wide") + 
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size = 15))

Alphaproteobacteria %>%
  ggplot(aes(x = Order, y = CountedGANTCAT / ExpectedGANTCAT)) +
  geom_boxplot() +
  geom_hline(yintercept = 1, linetype = "solid", color = "black") +
  theme_classic() +
  geom_point(data = Alphaproteobacteria %>% filter(Strain == "NA_1000"), 
             aes(x = Order, y = CountedGANTCAT / ExpectedGANTCAT, color = "C.crescentus NA1000"), size = 3) +
  geom_point(data = Alphaproteobacteria %>% filter(Strain == "2308"), 
             aes(x = Order, y = CountedGANTCAT / ExpectedGANTCAT, color= paste("Brucella 2308","Chromosome: ", Chromosomes)), size = 3) +
  ylab("C / E # of GANTC site (N = A,T)") +
  theme(legend.position = "none") +
  ggtitle("N = A or T") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size = 15))

Alphaproteobacteria %>%
  ggplot(aes(x = Order, y = CountedGANTCGC / ExpectedGANTCGC)) +
  geom_boxplot() +
  geom_hline(yintercept = 1, linetype = "solid", color = "black") +
  theme_classic() +
  geom_point(data = Alphaproteobacteria %>% filter(Strain == "NA_1000"), 
             aes(x = Order, y = CountedGANTCGC / ExpectedGANTCGC, color = "C.crescentus NA1000"), size = 3) +
  geom_point(data = Alphaproteobacteria %>% filter(Strain == "2308"), 
             aes(x = Order, y = CountedGANTCGC / ExpectedGANTCGC,  color= paste("Brucella 2308","Chromosome: ", Chromosomes)), size = 3) +
  ylab("C / E # of GANTC site (N = G,C)") +
  theme(legend.position = "none") +
  ggtitle("N = G or C") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size = 15))

Alphaproteobacteria %>%
  ggplot(aes(x = Order, y = CountedGANTCORF / ExpectedGANTCORF)) +
  geom_boxplot() +
  geom_hline(yintercept = 1, linetype = "solid", color = "black") +
  theme_classic() +
  geom_point(data = Alphaproteobacteria %>% filter(Strain == "NA_1000"), 
             aes(x = Order, y = CountedGANTCORF / ExpectedGANTCORF, color = "C.crescentus NA1000"), size = 3) +
  geom_point(data = Alphaproteobacteria %>% filter(Strain == "2308"), 
             aes(x = Order, y = CountedGANTCORF / ExpectedGANTCORF,  color= paste("Brucella 2308","Chromosome: ", Chromosomes)), size = 3) +
  ylab("C / E # of GANTC site in ORF") +
  theme(legend.position = "none") +
  ggtitle("Genic") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size = 15))

Alphaproteobacteria %>%
  ggplot(aes(x = Order, y = CountedGANTCIntergenic / ExpectedGANTCIntergenic)) +
  geom_boxplot() +
  geom_hline(yintercept = 1, linetype = "solid", color = "black") +
  theme_classic() +
  geom_point(data = Alphaproteobacteria %>% filter(Strain == "NA_1000"), 
             aes(x = Order, y = CountedGANTCIntergenic / ExpectedGANTCIntergenic, color = "C.crescentus NA1000"), size =3)+
  geom_point(data = Alphaproteobacteria %>% filter(Strain == "2308"), 
             aes(x = Order, y = CountedGANTCIntergenic / ExpectedGANTCIntergenic,  color= paste("Brucella 2308","Chromosome: ", Chromosomes)), size = 3) +
  ylab("C / E # of GANTC site in Intergenic") +
  theme(legend.position = "none") +
  ggtitle("Intergenic") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size = 15))

```